### PR TITLE
MOB-452 : add cta button to submit trades

### PR DESCRIPTION
**Pull request by mike-dydx**
## Links (dYdX Internal Use Only)
Linear Ticket: [MOB-452 : add cta button to submit trades](https://linear.app/dydx/issue/MOB-452/add-cta-button-to-submit-trades)

Figma Design: https://www.figma.com/file/mKevZOfE9nj6MZpiolKYW1/dYdX-%E2%80%BA-Mobile?type=design&node-id=5621-13681&mode=design&t=yQplL4xfSnUqF508-4

<br/>

## Description / Intuition
- adds cta button to tl/sp screen
- adds logic to display in-line errors with correct text
- adds logic to enable/disable the button based on validation errors
- hooks up new submit order action (note this is not working e2e because of v4 bug w/ collat checks)

<br/>

## Before/After Screenshots or Videos

| Before | After |
|--------|-------|
| <video src=""> | <video src="https://github.com/dydxprotocol/v4-native-ios/assets/3445394/be1e0bb7-7878-40cf-aa2b-ff48bd8dd151"> |

<br/>

### Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring or Technical Debt
- [ ] Documentation update
- [ ] Other (please describe: )

### CODE REVIEW

File: dydx/dydxPresenters/dydxPresenters/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewPresenter.swift [hunk 1]
```diff
@@ -69,17 +69,34 @@ private class dydxTakeProfitStopLossViewPresenter: HostedViewPresenter<dydxTakeP
...
+            .CombineLatest(
+                AbacusStateManager.shared.state.triggerOrdersInput,
+                AbacusStateManager.shared.state.validationErrors
+            )
             .compactMap { $0 }
-            .sink { [weak self] errors in
-                self?.updateValidationErrors(errors)
+            .sink { [weak self] triggerOrdersInput, errors in
+                #if DEBUG
+                if let triggerOrdersInput {
```
Comment (mike-dydx) [inline dydx/dydxPresenters/dydxPresenters/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewPresenter.swift:80]: will remove this in the future

File: dydx/dydxPresenters/dydxPresenters/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewPresenter.swift [hunk 2]
```diff
@@ -115,55 +132,63 @@ private class dydxTakeProfitStopLossViewPresenter: HostedViewPresenter<dydxTakeP
...
+            case "TRIGGER_MUST_BELOW_INDEX_PRICE":
+                print("mmm: \(error)")
+                if triggerOrdersInput?.stopLossOrder?.side == .sell {
+                    viewModel?.takeProfitStopLossInputAreaViewModel?.stopLossAlert = alert
+                }
+                if triggerOrdersInput?.takeProfitOrder?.side == .buy {
+                    viewModel?.takeProfitStopLossInputAreaViewModel?.takeProfitAlert = alert
+                }
+            default:
+                print("mmm: ", error.code)
```
Comment (mike-dydx) [inline dydx/dydxPresenters/dydxPresenters/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewPresenter.swift]: @ruixhuang  here is the error display logic

File: dydx/dydxPresenters/dydxPresenters/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewPresenter.swift (hunk 1 above)
Comment (ruixhuang) [reply dydx/dydxPresenters/dydxPresenters/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewPresenter.swift:80]: 🐞

File: dydx/dydxPresenters/dydxPresenters/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewPresenter.swift [hunk 3]
```diff
@@ -115,55 +132,70 @@ private class dydxTakeProfitStopLossViewPresenter: HostedViewPresenter<dydxTakeP
...
+        if let _ = triggerOrdersInput?.stopLossOrder?.price?.limitPrice?.doubleValue {
+            AbacusStateManager.shared.triggerOrders(input: Abacus.OrderType.stoplimit.rawValue, type: .stoplossordertype)
+        } else {
             AbacusStateManager.shared.triggerOrders(input: Abacus.OrderType.stopmarket.rawValue, type: .stoplossordertype)
         }
+
+        if let error = errors.first {
+            viewModel?.submissionReadiness = .fixErrors(cta: error.resources.action?.localized)
+        } else if triggerOrdersInput?.takeProfitOrder?.price?.triggerPrice?.doubleValue == nil
+            && triggerOrdersInput?.takeProfitOrder?.orderId == nil
```
Comment (ruixhuang) [inline dydx/dydxPresenters/dydxPresenters/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewPresenter.swift:192]: I am not doing this check on Android.  It seems to ready for submission if there is no ValidationError.

File: dydx/dydxPresenters/dydxPresenters/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewPresenter.swift [hunk 4]
```diff
@@ -115,55 +132,70 @@ private class dydxTakeProfitStopLossViewPresenter: HostedViewPresenter<dydxTakeP
...
+    private func update(triggerOrdersInput: TriggerOrdersInput?, errors: [ValidationError]) {
+
+        viewModel?.takeProfitStopLossInputAreaViewModel?.takeProfitAlert = nil
+        viewModel?.takeProfitStopLossInputAreaViewModel?.stopLossAlert = nil
+
+        for error in errors {
+            let alert = InlineAlertViewModel(.init(title: error.resources.title?.localized, body: error.resources.text?.localized, level: .error))
+            switch error.code {
+            case "TRIGGER_MUST_ABOVE_INDEX_PRICE":
+                print("mmm: \(error)")
```
Comment (ruixhuang) [inline dydx/dydxPresenters/dydxPresenters/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewPresenter.swift:144]: FYI, the "error.fields" should now contains the TriggerOrdersInputField where the error occurs..  It's probably easier to check for that.

Comment (mike-dydx) [reply dydx/dydxPresenters/dydxPresenters/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewPresenter.swift:144]: 👍  i will do that in a follow up PR

File: dydx/dydxViews/dydxViews/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewModel.swift [hunk 5]
```diff
@@ -61,24 +71,61 @@ public class dydxTakeProfitStopLossViewModel: PlatformViewModel {
         }
     }
 
+    private func createCta(parentStyle: ThemeStyle, styleKey: String?) -> AnyView? {
+        let buttonText: String
+        let buttonState: PlatformButtonState
+        switch submissionReadiness {
+        case .readyToSubmit:
+            buttonText = DataLocalizer.shared?.localize(path: "APP.TRADE.ADD_TRIGGERS", params: nil) ?? ""
```
Comment (ruixhuang) [inline dydx/dydxViews/dydxViews/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewModel.swift:79]: We'd also need an "Edit Trigger" in addition to "Add Trigger".

**Review by ruixhuang — APPROVED** [review]

File: dydx/dydxPresenters/dydxPresenters/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewPresenter.swift (hunk 3 above)
Comment (mike-dydx) [reply dydx/dydxPresenters/dydxPresenters/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewPresenter.swift:192]: went over this with mulan today, without this check, the "Add triggers" button is enabled even with empty input. I don't think that is great UX, what do you think?

We discussed adding something to abacus like `containsChanges` to denote whether that button should be enabled or not, even with no errors

File: dydx/dydxViews/dydxViews/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewModel.swift (hunk 5 above)
Comment (mike-dydx) [reply dydx/dydxViews/dydxViews/_v4/TakeProfitStopLoss/dydxTakeProfitStopLossViewModel.swift:79]: will add

**Pull request merged**